### PR TITLE
Introduce `MixedBitSet`

### PR DESCRIPTION
**Pull request by nnethercote**
`ChunkedBitSet` is good at avoiding excessive memory usage for programs with very large functgions where dataflow bitsets have very large domain sizes. But it's overly heavyweight for small bitsets, because any non-empty `ChunkedBitSet` takes up at least 256 bytes.

This PR introduces `MixedBitSet`, which is a simple bitset that uses `BitSet` for small/medium bitsets and `ChunkedBitSet` for large bitsets. It's a speed and memory usage win.

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by nnethercote** [comment]
After #133431 removed `HybridBitSet` it might seem strange to immediately introduce a new bitset. The rationale is:
- it's a perf win; and
- `MixedBitSet` is a lot simpler than `HybridBitSet`.

File: compiler/rustc_mir_transform/src/lint_tail_expr_drop_order.rs [hunk 1]
```diff
@@ -49,24 +49,24 @@ struct DropsReachable<'a, 'mir, 'tcx> {
...
+        let make_new_path_set = || Rc::new(RefCell::new(MixedBitSet::new_empty(move_set_size)));
 
         let data = &self.body.basic_blocks[block];
         let Some(terminator) = &data.terminator else { return };
-        // Given that we observe these dropped locals here at `block` so far,
-        // we will try to update the successor blocks.
-        // An occupied entry at `block` in `self.visited` signals that we have visited `block` before.
+        // Given that we observe these dropped locals here at `block` so far, we will try to update
+        // the successor blocks. An occupied entry at `block` in `self.visited` signals that we
+        // have visited `block` before.
```
Comment (Mark-Simulacrum) [inline compiler/rustc_mir_transform/src/lint_tail_expr_drop_order.rs:65]: Is reflowing a bunch of comments here intentional?

File: compiler/rustc_index/src/bit_set.rs [hunk 2]
```diff
@@ -1388,7 +1543,7 @@ where
     C: Idx,
 {
     num_columns: usize,
-    rows: IndexVec<R, Option<ChunkedBitSet<C>>>,
+    rows: IndexVec<R, Option<BitSet<C>>>,
```
Comment (Mark-Simulacrum) [inline compiler/rustc_index/src/bit_set.rs:1546]: Nit, stale comment above on Chunked.

**Review by Mark-Simulacrum — APPROVED** [review]
r=me with or without nits fixed

File: compiler/rustc_mir_transform/src/lint_tail_expr_drop_order.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_mir_transform/src/lint_tail_expr_drop_order.rs:65]: Yeah, I mentioned it in the commit log: "Also reflow some overlong comment lines in `lint_tail_expr_drop_order.rs`."

**Comment by nnethercote** [comment]
I addressed the comments.

@bors r=Mark-Simulacrum

**Comment by Mark-Simulacrum** [comment]
Overall positive, all primary benchmarks show improvement (or no change).

**Pull request merged**